src/queue.ml

```diff
@@ -125,12 +126,19 @@ module Make (B : Context.A_DISK) = struct
     in
     size ptr
 
-  let rec push_discarded ~quantity t =
+  let rec push_discarded ~quantity t bitset =
+    let rec free = function
+      | a :: b ->
+        let* () = Bitset.free_range bitset a in
+        free b
+      | [] -> Lwt_result.return ()
+    in
     match B.acquire_discarded () with
     | [] -> Lwt_result.return (t, quantity)
     | lst ->
       let* t = push_back_list t lst in
-      push_discarded ~quantity:(quantity + List.length lst) t
+      let* () = free lst in
```

Comment:
(leaving a note to remember that we need the bitset and the queue to be disjoint in the future)